tornado/web.py

```diff
@@ -2809,6 +2809,23 @@ def validate_absolute_path(self, root: str, absolute_path: str) -> Optional[str]
                 self.redirect(self.request.path + "/", permanent=True)
                 return None
             absolute_path = os.path.join(absolute_path, self.default_filename)
+        else:
+            canonical_path = os.path.normcase(os.path.normpath(absolute_path))
+            new_url = os.path.relpath(canonical_path, self.root).replace(os.path.sep, '/')
+            if os.path.join(self.root, self.path).replace('/', os.path.sep)\
+                    != canonical_path or not self.request.path.endswith(new_url):
+                if not new_url.startswith('/'):
+                    new_url = '/' + new_url
+                if self.request.path.endswith(new_url):
+                    path_prefix = self.request.path[:-len(new_url)]
+                elif urllib.parse.unquote(self.request.path).endswith(new_url):
+                    path_prefix = urllib.parse.unquote(self.request.path)[:-len(new_url)]
+                else:
+                    raise HTTPError(403, "could not replicate URL parsing for %s", self.path)
+                path_prefix = os.path.normpath(path_prefix).replace(os.path.sep, '/')
+                full_new_url = (path_prefix + new_url).replace(os.path.sep, '/')
+                self.redirect(full_new_url, permanent=True)
```

Comment:
Passing user input to `redirect` can be dangerous. I think we at least need an assertion that the url begins with a slash (so it's not an open redirector) and that it doesn't contain newlines or other special characters (there's another check for this deeper in tornado, but it wouldn't hurt to be explicit here).

Reply:
If `redirect` doesn't already do that, as I expected, yes.